### PR TITLE
Decouple data artifacts from Docker images

### PR DESCRIPTION
**Pull request by 3nprob**
This is still just a draft, feedback welcome.
All artifacts and data are stored under `./data`. To build, do `make data/London.photon`.

#### nominatim
- Remove image (using upstream directly now)

#### photon
- Add env var HEADWAY_NOMINATIM_FILE for source sql file path
- Add env var HEADWAY_PHOTON_LANGUAGES
- Don't process data as part of image build

#### Makefile
- Mount local filesystem instead of docker volumes in
  nominatim,planetiler,photon

## Breaking changes
A bit of previous build-time config is now read at runtime. In particular, ensure to set the web config for the nginx container (see `.env.example`)

TODO:
- [x] Align all Makefile targets
- [x] Wire up docker-compose.yaml
- [x] clean
- [ ] End-to-end deployment test
- [x] Allow specifying custom path instead of `./data` (and possibly split up further into subdirectories rather than mounting it directly?)
- [x] Update docs

### CODE REVIEW

**Comment by 3nprob** [comment]
While one hand it's nice that each step is just reading and writing files, it's a bit redundant to spin up ephemeral psql servers in separate steps and shuffling sql data between them.

Maybe at some point it would worth looking into using a shared instance and skip the pgdump/import? I can see why it makes sense to keep that as it is until someone raises it as a practical issue, though.

**Comment by ellenhp** [comment]
We're working on essentially the same PR apparently. I started trying to add OTP and got sidetracked by trying to remove some of the awful hacks. Do you anticipate having enough free time to finish this in the next few days? If so I can pivot to working on OTP/transit stuff and mop up some frontend bugs.

**Comment by ellenhp** [comment]
> While one hand it's ergonomic and nice that each step is just reading and writing files, it's a bit redundant to spin up ephemeral psql servers in separate steps and shuffling sql files between them.

I think it makes a lot of sense to store the nominatim artifact as whatever comes out of pg_dump (probably the tar format?) because it seems like a pg_restore can be very fast if you use specific config options. What I'm imagining is the nominatim container in docker-compose will just mount the data directory and then a startup script will pg_restore from data/nominatim.tar or whatever we call it, then begin serving.

I'm not really sure what to do about photon though. Maybe just a straight tarball of the elasticsearch data directory?

**Comment by ellenhp** [comment]
That's definitely a hack and wouldn't scale well to larger deployments though, but it would be less finicky than PostGIS on a persistent volume.

**Comment by 3nprob** [comment]
> We're working on essentially the same PR apparently. I started trying to add OTP and got sidetracked by trying to remove some of the awful hacks. Do you anticipate having enough free time to finish this in the next few days? If so I can pivot to working on OTP/transit stuff and mop up some frontend bugs.

I can give it a shot! Would be helpful if you take a quick look at the changes so far to see if there's already something you'd rather see differently etc :)

**Comment by 3nprob** [comment]
> What I'm imagining is the nominatim container in docker-compose will just mount the data directory and then a startup script will pg_restore from data/nominatim.tar or whatever we call it, then begin serving.

So basically where I'm going here is that the tar step is skipped but otherwise like that; nominatim does pg_dump in one place and photon imports it straight from the same file.

File: Makefile [hunk 1]
```diff
@@ -38,77 +38,53 @@ list:
...
 	grep "$(basename $@):" gtfs/bboxes.csv > $@
 	perl -i.bak -pe 's/$(basename $@)://' $@
 
 %.mbtiles: %.osm.pbf
-	@echo "Building MBTiles $(basename $@)"
-	mkdir -p ./.tmp_mbtiles
-	cp $(basename $@).osm.pbf ./.tmp_mbtiles/data.osm.pbf
-	docker volume create headway_mbtiles_build || echo "Volume already exists"
+	@echo "Building MBTiles $(notdir $(basename $@))"
 	docker build ./mbtiles/bootstrap --tag headway_mbtiles_bootstrap
```
Comment (ellenhp) [inline Makefile:50]: There's another easy win here, we could just untar sources.tar into the data directory. The only reason it was a container before was that gave it access to a volume, but if the volume is a host directory the makefile can do this and we can remove the bootstrap image.

File: geocoder/photon/import_from_dump.sh [hunk 2]
```diff
@@ -12,8 +15,8 @@ sudo -E -u postgres psql postgres -c "DROP DATABASE IF EXISTS nominatim"
 
 sudo -E -u postgres psql postgres -c "CREATE DATABASE nominatim"
 
-sudo -E -u postgres psql nominatim < /nominatim_data/nominatim
+sudo -E -u postgres psql nominatim < ${HEADWAY_NOMINATIM_FILE}
 
-sudo -u photon /bin/sh -c 'java -jar photon.jar -nominatim-import -host localhost -port 5432 -database nominatim -user nominatim -password password1 -languages es,fr,de,en'
+sudo -u photon /bin/sh -c "java -jar /photon/photon.jar -nominatim-import -host localhost -port 5432 -database nominatim -user nominatim -password password1 -languages ${HEADWAY_PHOTON_LANGUAGES}"
```
Comment (ellenhp) [inline geocoder/photon/import_from_dump.sh:20]: Nice, good catch. I've tried to be good about adding more than just the languages I speak but making this a build option is ideal. :)

**Review by ellenhp — COMMENTED** [review]
Yeah this looks fine so far! You're definitely making quicker progress than I have been and it seems like you know your way around makefile string manipulation stuff a bit better. not-dir was news to me.

**Comment by ellenhp** [comment]
> > What I'm imagining is the nominatim container in docker-compose will just mount the data directory and then a startup script will pg_restore from data/nominatim.tar or whatever we call it, then begin serving.
> 
> So basically where I'm going here is that the tar step is skipped but otherwise like that; nominatim does pg_dump in one place and photon imports it straight from the same file.

The "fun" part of decoupling these is that now photon and nominatim both have two cases they need to support instead of one. Initial import/bootstrapping and then also the configuration they'll be in when they're started in compose or k8s and need to serve traffic asap. It'll be worth it to decouple them from data, but there are a lot of considerations about how to store data that I haven't fully thought through.

**Comment by 3nprob** [comment]
> The "fun" part of decoupling these is that now photon and nominatim both have two cases they need to support instead of one. Initial import/bootstrapping and then also the configuration they'll be in when they're started in compose or k8s and need to serve traffic asap. It'll be worth it to decouple them from data, but there are a lot of considerations about how to store data that I haven't fully thought through.

In general,as long as the two modes are distinct and have different entrypoints (ie if in "run" mode and necessary bootstrapping is not done, exit with error and let the operator sort that out instead of trying to be "smart")

photon looks straightforward from what I can tell (looks like it generates an ES db from the sql?)

The exception in the current state of this PR is nomatim and its built-in postgres db, which currently is regenerating on each container start. For now I guess the easy solution for this would be to persist that in a volume mount as well

**Review by ellenhp — COMMENTED** [review]
This caused a massive (50x or so) performance regression during nominatim index generation and I can't for the life of me figure out why. I'm going to continue looking at that today.

**Comment by ellenhp** [comment]
It looks like the performance tuning in config.sh here isn't actually being applied: https://github.com/mediagis/nominatim-docker/blob/master/4.0/config.sh

**Comment by ellenhp** [comment]
I'm going to PR some changes into your branch rather than keep making comments I think. Should be up within an hour or two.

**Comment by ellenhp** [comment]
See: #18

**Comment by 3nprob** [comment]
> I also want to apologize for the whiplash here

All good, two steps forward and one back is still forward.

Following up in https://github.com/3nprob/headway/pull/1, which superseded #18 and #19.

**Pull request merged**